File: backend/src/main/java/org/bootstrapbugz/api/auth/validator/UsernameOrEmail.java
package org.bootstrapbugz.api.auth.validator;

import static java.lang.annotation.ElementType.ANNOTATION_TYPE;
import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.ElementType.TYPE;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import org.bootstrapbugz.api.auth.validator.impl.UsernameOrEmailImpl;

@Target({TYPE, FIELD, ANNOTATION_TYPE})
@Retention(RUNTIME)
@Constraint(validatedBy = UsernameOrEmailImpl.class)
@Documented
public @interface UsernameOrEmail {
  String message() default "";

  Class<?>[] groups() default {};

  Class<? extends Payload>[] payload() default {};
}

File: backend/spring-boot-kotlin/src/main/kotlin/org/bootstrapbugz/api/shared/validator/UsernameOrEmail.kt
package org.bootstrapbugz.api.shared.validator

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass
import org.bootstrapbugz.api.shared.validator.impl.UsernameOrEmailImpl

@Target(AnnotationTarget.CLASS, AnnotationTarget.FIELD, AnnotationTarget.ANNOTATION_CLASS)
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [UsernameOrEmailImpl::class])
@MustBeDocumented
annotation class UsernameOrEmail(
  val message: String = "",
  val groups: Array<KClass<*>> = [],
  val payload: Array<KClass<out Payload>> = []
)
